rse-py-scripting.Rmd

```diff
@@ -0,0 +1,143 @@
+# Command line programs {#rse-py-scripting}
+
+## Questions {#rse-py-scripting-questions}
+
+```{r, child="questions/rse-py-scripting.md"}
+```
+
+
+## Introduction {#rse-py-scripting-intro}
+
+The Jupyter Notebook, PyCharm and other interactive tools
+are great for prototyping code and exploring data,
+but in many cases you will ultimately want to take the code you've developed
+using those tools and apply it to thousands of data files, 
+run it with many different parameters,
+and/or combine it with other programs in a data analysis pipeline. 
+In order to do these things,
+it is often necessary/preferable to turn the code you've developed into a program 
+that works like any other Unix command-line tool.
+
+In this chapter we take the code we developed earlier (TODO: Link)
+and turn it into a Python script (i.e. a command line program written in Python).
+
+
+## A template script {#rse-py-scripting-template}
+
+In general, the first thing that gets added to any Python script is the following:
+
+```python
+if __name__ == '__main__':
+    main()
+```
+
+The reason we need these two lines of code
+is that running a Python script in bash is very similar to importing that file in Python. 
+The biggest difference is that we don’t expect anything to happen when we import a file, 
+whereas when running a script we expect to see some output
+(e.g. an output file, figure and/or some text printed to the screen).
+
+The `__name__` variable exists to handle these two situations.
+When you import a Python file `__name__` is set to the name of that file
+(e.g. when importing script.py, `__name__` is `script`),
+but when running a script in bash `__name__` is always set to `__main__`.
+The convention is to call the function that produces the output `main()`,
+but you can call it whatever you like.
+
+The next thing you'll need is a library to parse the command line for input arguments.
+The most widely used option is 
+[argparse](https://docs.python.org/3/library/argparse.html).
+
+Putting those together,
+here's a basic python script for us to start with:
+
+```shell
+$ cat script_template.py
```

Comment:
Do we need this code block? Can we say "Create the file `script_template.py`" instead?